core/src/main/java/io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java

```diff
@@ -0,0 +1,605 @@
+        public void toTop() {
+            baseCursor.toTop();
+            gapFilling = false;
+            baseRecord = null;
+            fillOffset = 0;
+        }
+
+
+        private Function getFillFunction(int col) {
+            if (col == timestampIndex) {
+                return TimestampConstant.newInstance(nextBucket);
+            }
+
+            return values.getQuick(col < timestampIndex ? col : col - 1);
+        }
+
+        private void initBounds(Function from, Function to) {
+            if (from != TimestampConstant.NULL
+                    && to != TimestampConstant.NULL) {
+                rangeBound = RANGE_FULLY_BOUND;
+                return;
+            }
+
+            if (from != TimestampConstant.NULL) {
+                rangeBound = RANGE_LOWER_BOUND;
+                return;
+            }
+
+            if (to != TimestampConstant.NULL) {
+                rangeBound = RANGE_UPPER_BOUND;
+                return;
+            }
+
+            if (from == TimestampConstant.NULL && to == TimestampConstant.NULL) {
+                rangeBound = RANGE_UNBOUNDED;
+            }
+        }
+
+
+        private void initTimestamps(Function from, Function to, CharSequence stride) throws SqlException {
+            if (from != TimestampConstant.NULL) {
+                fromTimestamp = from.getTimestamp(null);
+            }
+
+            if (to != TimestampConstant.NULL) {
+                toTimestamp = to.getTimestamp(null);
+            }
+
+            timestampSampler = TimestampSamplerFactory.getInstance(stride, 0);
+            timestampSampler.setStart(fromTimestamp);
+
+            nextBucket = fromTimestamp;
+
+            minTimestamp = from == TimestampConstant.NULL ? Long.MAX_VALUE : fromTimestamp;
+            maxTimestamp = to == TimestampConstant.NULL ? Long.MIN_VALUE : toTimestamp;
+        }
+
+        private void moveToNextBucket() {
+            fillOffset++;
+            nextBucket = timestampSampler.nextTimestamp(nextBucket);
+        }
+
+        private boolean notAtEndOfBitset() {
+            if (rangeBound == RANGE_LOWER_BOUND || rangeBound == RANGE_UNBOUNDED) {
+                return fillOffset < timestampSampler.bucketIndex(maxTimestamp);
+            } else {
+                return fillOffset <= timestampSampler.bucketIndex(maxTimestamp);
+            }
+
+        }
+
+        private void of(RecordCursor baseCursor, SqlExecutionCircuitBreaker circuitBreaker, Function from, Function to, CharSequence stride, ObjList<Function> values, int timestampIndex) throws SqlException {
+            assert from != null;
+            assert to != null;
+            this.baseCursor = baseCursor;
+            this.circuitBreaker = circuitBreaker;
+            this.timestampIndex = timestampIndex;
+            this.values = values;
+            initTimestamps(from, to, stride);
+            presentRecords = new BitSet(to != TimestampConstant.NULL ? timestampSampler.bucketIndex(toTimestamp) : 64 * 8);
```

Comment:
nit: might be a good idea to extract 64*8 to a constant and put a comment why exactly this value was chosen.